app/src/main/java/com/example/javaopencommerce/order/OrderIntegrityValidator.java

```diff
@@ -0,0 +1,135 @@
+
+  public Uni<Void> validateOrder(OrderDetailsDto order) {
+    orderInaccuracies = new ArrayList<>();
+    validateOrderBasics(order);
+    validateOrderState(order);
+    Uni<List<ItemDto>> currentItems = fetchCurrentItems(order.getCard().getProducts());
+    currentItems
+        .onItem().invoke(items -> validateOrderValue(items, order.getCard()))
+        .onItem().invoke(items -> validateStocks(items, order.getCard().getProducts()))
+        .await().indefinitely();
+
+    if (!orderInaccuracies.isEmpty()) {
+      throw new OrderValidationException(orderInaccuracies);
+    }
+
+    return Uni.createFrom().voidItem();
+  }
+
+  private Uni<List<ItemDto>> fetchCurrentItems(List<ProductDto> products) {
+    List<Long> itemIds = products.stream()
+        .filter(Objects::nonNull)
+        .map(ProductDto::getItem)
+        .filter(Objects::nonNull)
+        .map(ItemDto::getId)
+        .collect(toUnmodifiableList());
+    return itemQueryRepository.getItemsByIdList(itemIds);
+  }
+
+  private void validateOrderBasics(OrderDetailsDto order) {
+    if (order.getAddressId() == null) {
+      orderInaccuracies.add(new AddressNotProvidedValidationException());
+    }
+  }
+
+  private void validateOrderState(OrderDetailsDto order) {
+    if (order.getOrderStatus() == null || !order.getOrderStatus().equals("NEW")) {
+      orderInaccuracies.add(new OrderStateValidationException(
+          format("Wrong/Illegal order status: %s", order.getOrderStatus())));
+    }
+
+    if (order.getPaymentStatus() == null || !order.getPaymentStatus().equals("BEFORE_PAYMENT")) {
+      orderInaccuracies.add(new OrderStateValidationException(
+          format("Wrong/Illegal payment status: %s", order.getPaymentStatus())));
+    }
+
+    try {
+      PaymentMethod.valueOf(order.getPaymentMethod());
+    } catch (IllegalArgumentException e) {
+      orderInaccuracies.add(new OrderStateValidationException(
+          format("Wrong/Illegal payment method: %s", order.getPaymentMethod())));
+    }
+  }
+
+  private void validateOrderValue(List<ItemDto> currentItems, CardDto card) {
+    Map<Long, BigDecimal> currentItemPrices = currentItems.stream()
+        .filter(Objects::nonNull)
+        .collect(toMap(ItemDto::getId, ItemDto::getValueGross));
+
+    BigDecimal currentOrderValue = calculateCurrentOrderValue(card.getProducts(),
+        currentItemPrices);
+
+    if (!currentOrderValue.equals(card.getCardValueGross())) {
+      orderInaccuracies.add(
+          new OrderValueNotMatchingValidationException(card.getCardValueGross(),
+              currentOrderValue));
+    }
+  }
+
+  private BigDecimal calculateCurrentOrderValue(List<ProductDto> orderedProducts,
+      Map<Long, BigDecimal> currentPrices) {
+
+    return orderedProducts.stream()
+        .map(product ->
+            ofNullable(currentPrices.get(product.getItem().getId()))
+                .orElseThrow(
+                    () -> new RuntimeException("Some custom exception for validation error"))
```

Comment:
Runtime :(

Reply:
Hmm switched to the OrderValidationException with custom text, should be fine for now